doc/BACKUP.md: Document backup strategies for `lightningd`.

ZmnSCPxj opened the pull request:
Requested by @d4amenace here: https://github.com/ElementsProject/lightning/issues/4200#issuecomment-727642350

@cdecker to finish up the document re: PostgreSQL.  I remember somebody made a medium article on C-Lightning PostgreSQL but I did not save the link...

Changelog-None

jsarenik commented:
Let me link also https://github.com/ElementsProject/lightning/issues/4181 for reference.

Thank you @ZmnSCPxj for working on it!

ZmnSCPxj commented:
Clarified location of `$LIGHTNINGDIR`.

ZmnSCPxj commented:
Corrected misspelling of `lighningd`.

jsarenik commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,312 @@
+
+You can also use any automated backup system as long as it includes the
+`lightningd.sqlite3` file (and optionally `hsm_secret`, but note that
+as a secret key, hackers getting a copy of your backups may allow them
```
hackers? Not sure we should jump on the train calling thieves hackers.

http://www.catb.org/jargon/html/index.html

jsarenik commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,312 @@
+
+However, again, note that a "no backups #reckless" strategy leads to
+*definite* loss of funds, so you should still prefer this strategy rather
+than not have **any** backups at all.
```
Emphasis on word **any** is a bit misleading. There is no comparision (with other emphasized word nearby). I would prefer something like "so you should still prefer this strategy rather than having no backups at all."

ZmnSCPxj replied:
"you should still prefer **this** strategy rather than have **no** backups at all." reads better?

jsarenik replied:
Sure. That works well.

jsarenik commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,312 @@
+than not have **any** backups at all.
+
+Even if you have the better options below, you might still want to do
+this as a fallback, as long as you ensure:
```
as long as you: (and see below)

jsarenik commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,312 @@
+Even if you have the better options below, you might still want to do
+this as a fallback, as long as you ensure:
+
+* Attempt to recover using the other backup options below first.
```
This is good as it is. Actually it does not go well when there is the "ensure" word above.

jsarenik commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,312 @@
+
+* Attempt to recover using the other backup options below first.
+  Any one of them will be better than this backup option.
+* Recovering by this method ***MUST*** always be the ***last*** resort.
```
Change to something like "Use this method **ONLY** as a ***last*** resort" (goes well with "as long as you:" above)

ZmnSCPxj replied:
Went with "Recover by this method **ONLY** as a ***last*** resort"

jsarenik commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,312 @@
+BTRFS would probably work better if you were purchasing an entire set
+of new storage devices to set up a new node.
+
+On BSD you can use a ZFS RAID-Z setup, which is probably better than BTRFS
```
The general advice (like first paragraph in this section also) could be moved together, BTRFS could be turned into its own sub-section so that it is optically skippable (sp?).

BTW this review comes from a point of view of a non-native speaker. So take it with a grain of salt.

ZmnSCPxj replied:
Okay.  I guess we can list software methods of RAID-1, such as:

* `mdadm`
* BTRFS
* ZFS

jsarenik commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,312 @@
+you to still recover your Lightning funds.
+
+However, if instead you are just replicating the database on another
+storage device on a single location, you remain vulnerable to disasters
```
in a single location?

ZmnSCPxj replied:
Right, "in" is better than "on" here.

jsarenik reviewed (changes requested):
Wonderful documentation (except the BTRFS ad :-P)
Thank you!

ZmnSCPxj commented:
Updated as per @jsarenik feedback.

cdecker commented:
I think you're referring to @gabridome's excellent tutorial here: https://github.com/gabridome/docs/blob/master/c-lightning_with_postgresql_reliability.md

ZmnSCPxj commented:
Add `autodefrag` to BTRFS instructions, minor wording tweak, expand "PostgreSQL" section.

darosior commented:
Would be worse to link it from https://lightning.readthedocs.io/FAQ.html#how-to-backup-my-wallet i think

ZmnSCPxj commented:
Mention the SQLITE3 backup API, add link from FAQ.md to BACKUP.md

ZmnSCPxj commented:
Mention using the `devid` in BTRFS.

ZmnSCPxj commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,391 @@
+
+This creates a consistent snapshot of the database, sampled in a
+transaction, that is assured to be openable later by `sqlite3`.
+The operation of the `lightningd` process will be paused while `sqlite3`
```
***THIS IS FACTUALLY INCORRECT***.

I did a stress test where a program continuously makes many small transactions to update a database.  When you do something like `.backup 'backup.sqlite3'` or `VACUUM INTO `backup.sqlite3';` in a *separate* `sqlite3`, then either the main program gets an `SQLITE_BUSY`, ***or*** the separate `sqlite3` returns a "Database is locked" error.

I suspect that an `SQLITE_BUSY` will cause `lightningd` to crash, so this is probably strongly not recommended to use a separate process to back up.

We can do:

* Expose an `sqlite3_backup` command which performs the `VACUUM INTO` query inside `lightningd`.
* Call into `sqlite3_busy_timeout` with a "reasonable" large value, say 5000 (5 seconds). (Equivalently, call `PRAGMA busy_timeout = 5000;` query at the same time we do `PRAGMA foreign_keys=on;`).  Then, even if the `.backup` or `VACUUM INTO` takes up to 5 seconds to back up, `lightningd` will just wait.

The latter solution is easier but there is always the possibility that the backup process will take more than whatever timeout we select.  The earlier solution is more reliable and there is no timeout, but note that if say the target location is slow (e.g. an NFS mount) then `lightningd` will suspend indefinitely.

Thoughts?

jsarenik replied:
Thank you @ZmnSCPxj for ideas and investigating possible issues with running backup in separate `sqlite3` process!

Exposing the `sqlite3_backup` command would be great! It sounds like a very nice and clear way to do backup.

ZmnSCPxj commented:
Remove the factually incorrect text mentioned in https://github.com/ElementsProject/lightning/pull/4207#discussion_r528206037

ZmnSCPxj commented:
Teach how to quickly check if a backup database is corrupted.

MrManPew commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,400 @@
+
+This is the least desirable backup strategy, as it *can* lead to loss
+of all in-channel funds if you use it.
+However, having *no* backup strategy at all *will* lead to loss of all
```
Not sure it is true, as eventually the other user will force-close the channels and you will get the BTC back to your address? right?

ZmnSCPxj replied:
If the peer does not have `option_static_remotekey` then the address it goes to is not on the path derived from `hsm_secret`, but requires the derivation keys stored in the db.  Though lots of nodes today already support `option_static_remotekey`.

ZmnSCPxj replied:
You also need to wait for the other side to actually force close, and that could take a while.

cdecker replied:
No, if a channel was opened without the static key option the payout address is dependent on the last state (address is tweaked with the channel state), and the wallet will not be able to recognize the address as its own.

Worse: relying on the peer to do the right thing might expose you to fraud if they decide they prefer an older state where they had a bigger share of the channel capacity.

MrManPew commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,400 @@
+lose all funds in open channels.
+
+However, again, note that a "no backups #reckless" strategy leads to
+*definite* loss of funds, so you should still prefer *this* strategy rather
```
again, what if the other user close the channels? wouldn't you get the money back to your wallet?

cdecker replied:
Just to add a bit more complexity: if we use `option_upfront_shutdown_script` we may actually get the funds back to our wallet if the peer closes, because that removes the tweak to the their_unilateral/to_us transaction.

MrManPew reviewed (commented):
Good info

ZmnSCPxj commented:
Okay, we can actually implement an sqlite3-backup-snapshot in a plugin, I think.

Basically, `lightningd` performs transactions this way:

* `BEGIN;` to the sqlite3 connection.
* Send queries to the connection.  Write queries are stored in a log maintained by the `db_write` hook.
* Just before committing, trigger the `db_write` hook.
  * If it returns something other than `{"result":"continue"}`, exit the entire program, which implicitly rolls back the transaction.
* `COMMIT;` to the sqlite3 connection.

If so, a plugin that hooks into `db_write` can delay the `COMMIT;` query by the simple act of not responding immediately.  And the write(s) to the database file are triggered by `COMMIT;`; without the `COMMIT;` everything is in the `-journal` file.

So it looks to me that we can have a C plugin which hooks into `db_write`, and the `db_write` handler just responds with `{"result":"continue"}`, and offers a `sqlite3-snapshot` command.  The handler of the `sqlite3-snapshot` just uses synchronous code (which implicitly causes any `db_write` hook to block) to copy the `lightningd.sqlite3` file elsewhere.  After it does the copy it responds and unblocks to the mainloop, letting any pending `db_write` through (it could "simply" unblock after the copy and do `VACUUM` on the copy, possibly also deleting entries in non-vital tables like invoices, and non-pending payments to reduce the size of the snapshot).  The plugin can check if the `lightningd` is on `sqlite3` by checking `wallet` in `listconfigs`  I probably need to write some stress-test experimental code to simulate this.

In WAL mode we need to copy the `-wal` file as well.  It is possible to set your `lightningd.sqlite3` to WAL mode by stopping `lightningd`, then `sqlite3 lightningd.sqlite3` and issue `PRAGMA journal_mode=WAL`, then `.exit` the `sqlite3` and restarting `lightningd`.  You can revert with `PRAGMA journal_mode=DELETE;`

ZmnSCPxj commented:
> Okay, we can actually implement an sqlite3-backup-snapshot in a plugin, I think.

Not actually, because the `doc/PLUGINS.md` specifically says that a plugin that hooks into `db_write` should not handle any other hooks or commands.  Since we have this in our doc, we should really not violate it ourselves, as otherwise future contributors might provide code that assumes no decent plugin will handle `db_write` and another command.  So it would still be best to either add a decent busy timeout to our db connection so a third-party can snapshot the file in a separate `sqlite3` connection, or implement something inside the `lightningd` process to do the snapshot.

d4amenace commented:
> > ....... or implement something inside the `lightningd` process to do the snapshot.

This actually made sense to me.  There could be a configurable timeout period in the conf file to denote the period of snapshots.  At each interval writing would be paused for the db so that corrupted snapshots would be avoided.  There would be a standardized snapshot folder in the users .lighting folder which would be the most all-encompassing solutions of all the implementations.

ZmnSCPxj commented:
> There could be a configurable timeout period in the conf file to denote the period of snapshots.

All that `lightningd` has to expose at the most basic level would be the ability to make *one* snapshot at the current time.  Regular snapshots like what you want can be done by a separate program, xref. `autocleaninvoices` which has the scheduled invoice cleaning in a plugin, not in `lightningd` (`lightningd` just exposes a command to clean old invoices using a single db query, the auto-cleaning is done by the plugin).

In particular, such a regular snapshotting would be ***vastly*** inferior to `db_write`:

* During times when there are lots of updates to your channel state, your regular snapshotting would miss them.  If the node crashes at that point, you risk losing funds.
* During times when there is nothing happening, your regular snapshotting will copy files over and over, wasting space and time on substantially the same data.
  * If you want something "smart" like say "oh just save the changes between this snapshot and the previous one", ***that is what `db_write` already is***, *plus* it triggers at each update, so just use `db_write`, via the existing `backup.py` plugin, already.

Regular snapshotting of the db say once or twice a day would be a good "backup of a backup" --- your primary backup should still be a `db_write` plugin OR a PostgreSQL cluster OR at the absolute bare minimum some kind of RAID filesystem setup, and preferably all of them.  Recovering from a snapshot would be the sort of thing you would do as a last resort, only-use-this-if-Godzilla-attacks recovery in case all your *other* backups are broken.  And you would want that regular snapshotting to, say, be uploaded somewhere *else*, like encrypted on some cloud server or at least on a computer you own located in the house of a friend.  Since that probably involves some kind of custom arrangement, it would not be easy to standardize such an interface inside `lightningd`; so a single-shot snapshot would be enough in C-Lightning, I think, and you build your custom arrangement on top with `crontab`.

gabridome commented:
> Requested by @d4amenace here: [#4200 (comment)](https://github.com/ElementsProject/lightning/issues/4200#issuecomment-727642350)
> 
> @cdecker to finish up the document re: PostgreSQL. I remember somebody made a medium article on C-Lightning PostgreSQL but I did not save the link...

Sorry for the late reply. https://github.com/gabridome/docs/blob/master/c-lightning_with_postgresql_reliability.md

Available for any question, comment or help.

ZmnSCPxj commented:
@gabridome does the PostgreSQL section I made make sense?

gabridome commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,400 @@
+following command gives you a path:
+
+    pg_config --includedir
+
```
I didn't know it. When I built it, everything worked out of the box. Pretty useful to know. 
I probably had already the library installed.

ZmnSCPxj replied:
Okay, maybe the Lightning+PostgreSQL can be a separate PR?

gabridome replied:
The document you mean?
Something like `doc/postgresQL.md`?

Anyway This part in the backup strategies seems pretty right to me. At least to mention a more extended part in the doc directory.  @fiatjaf Has thought about that?

ZmnSCPxj replied:
Yes, the document I mean.

gabridome commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,400 @@
+You should use the same PostgreSQL version of `libpq-dev` as what you run
+on your cluster, which probably means running the same distribution on
+your cluster.
+
```
Personally I have added on the cluster machines the Postgres .deb repository in this way I can keep the versions aligned.
I didn't encounter any problem but maybe it is not wise to exit from the Postgres version installed with the Debian distribution. What you guys think?

ZmnSCPxj replied:
I would have used the same distro throughout just to reduce inter-version problems, which is why I suggested the above.  Mixing deb repositories has gotten me into trouble before when I wanted to upgrade my OS (but upgrading an OS is always fraught, sometimes it is best to just have two small OS partitions and alternate installing OS's between them rather than upgrading an existing install...)

gabridome commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,400 @@
+(though you should probably do some more double-checking and tire-kicking
+in the "Connect to the database" stage you resume at, such as checking if
+`listpeers` still lists the same channels as you had, and so on).
+
```
Wise suggestion IMO.

gabridome commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,400 @@
+Debian Testing ("bullseye") uses PostgreSQL 13.0 as of this writing.
+PostgreSQL 12 had a non-trivial change in the way the restore operation is
+done for replication.
+You should use the same PostgreSQL version of `libpq-dev` as what you run
```
I encountered the problem myself. I would add to always read the official synchronous replication PostgresQL guide for the **specific version** you are using.

ZmnSCPxj replied:
Okay, I will add that.

cdecker replied:
Agreed, there is little point in replicating the official doc here, it may just end up with stale information, requiring updates from us. If there is an authoritative source, link to it.

gabridome commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,400 @@
+[guide by @gabridome][gabridomeguide].
+
+[gabridomeguide]: https://github.com/gabridome/docs/blob/master/c-lightning_with_postgresql_reliability.md
+
```
The guide is not maintained as the versions of Postgres evolve. Please always check also the Postgres Guide about synchronous replication for your specific version.

gabridome commented:
> @gabridome does the PostgreSQL section I made make sense?

Absolutely. Thank you for mentioning my guide.

ZmnSCPxj commented:
Rebased, emphasized checking your PostgreSQL version, explicated more about tire-kicking after SQLITE3->PG conversion.

ZmnSCPxj commented:
No ACKs?  People seem to want this.  Is there any issue or quibble that prevents this from being merged?

gabridome commented:
> No ACKs? People seem to want this. Is there any issue or quibble that prevents this from being merged?

FWIW conceptACK.

I plaude loudly this incredible work on the most important topic of the LN that has been delayed for too long.

I frankly hope that these strategies will become more and more user friendly as the time passes. Until then, the paradox is that the ones less aware of the problem, are the one more in need of a good solution...

gabridome commented:
Hi,
there's a new possibility for the backup of `hsm_secret` file.
I find it useful because it converts to xpriv format, so you can import into one descriptor based wallet in Bitcoin Core or wherever xpriv are accepted.
https://github.com/domegabri/lightning-secret

ZmnSCPxj commented:
@gabridome how do we recover the `hsm_secret` from the xprv?  The program seems to only convert from `hsm_secret` to xprv.

gabridome commented:
Mmh... You're right. Gotta check.

On Thu, Dec 10, 2020, 11:17 PM ZmnSCPxj, ZmnSCPxj jxPCSmnZ <
notifications@github.com> wrote:

> @gabridome <https://github.com/gabridome> how do we recover the hsm_secret
> from the xprv? The program seems to only convert from hsm_secret to xprv.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ElementsProject/lightning/pull/4207#issuecomment-742834832>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADQQ3VOPM5CKDMEOSXWTCLSUFCGNANCNFSM4TYRZNBQ>
> .
>

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+in-channel activity*.
+This is in contrast to onchain payments, where you have to tell everyone
+about each and every payment and have it recorded on the blockchain,
+leading to scaling problems (you have to push data at everyone) and
```
```suggestion
leading to scaling problems (you have to push data to everyone, everyone needs to verify every transaction) and
```

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+
+But in Lightning, since *you* are the only one storing all your
+financial information, you ***cannot*** recover this financial
+information anywhere else.
```
```suggestion
information from anywhere else.
```

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+financial information, you ***cannot*** recover this financial
+information anywhere else.
+
+This means that on Lightning, you **have to** responsibly back up your
```
```suggestion
This means that on Lightning, **you have to** responsibly back up your
```
:wink:

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+You need a copy of the `hsm_secret` file regardless of whatever backup
+strategy you use.
+This is just 32 bytes, and you can do something like the below and
+write the hexadecimal digits a few times on a piece of paper:
```
Maybe mention that this file does not change after its creation, so backing it up when first setting up the node is sufficient and no ongoing backup is required, in contrast to the DB which requires continuous backups.

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+To recover in-channel funds, you need to use one or more of the other
+backup strategies below.
+
+Database File Backups
+---------------------
```
IMHO the methods should be sorted in descending order of desirability, putting this type first gives absolutely the wrong impression.

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+* Channels with peers that do not support `option_dataloss_protect`.
+  * Most nodes on the network already support `option_dataloss_protect`
+    as of November 2020.
+  * If the peer does not support this option, then the entire channel
+    funds will be revoked by the peer.
```
Technically not true: the peer may be well-behaved, and receive back his share of the funds, but the user's funds will be lost forever into an unrecoverable address, due to the tweak.

Even if the peer is not well-behaved, it can at most grab the funds that belonged to it some time in the past, not all funds. The peer would then go back to that point in time, grab it's now bigger share, and still sinkhole the user's share.

ZmnSCPxj replied:
How about "up to the entire channel funds can be revoked by the peer"?

ZmnSCPxj replied:
No, wait, this is true if the `option_dataloss_protect` is *not* supported, which is what qualifies the statement.  Without `option_dataloss_protect` we would get confused with the `reestablish_connection` and try to broadcast our obsolete revoked commitment transaction, and the peer will react by revoking the revoked commitment transaction.

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+    as of November 2020.
+  * If the peer does not support this option, then the entire channel
+    funds will be revoked by the peer.
+  * Specially-coded peers can pretend to support this option, but end
+    up taking the entire channel funds (or otherwise more than what it
+    should have) when you try to recover.
+    Standard releases of Lightning Network node software do not do this
+    (they just support this option according to the standard) but note
+    that anyone can modify their node software, especially with our
+    well-documented open-source code.
```
Doesn't change the behavior above: the peer can prevent the user from recovering its share of the funds by not sharing the commitment point.

cdecker replied:
Generally speaking I'd reduce the level of detail here slightly, to "if the peer is non-cooperative you won't be able to recover your funds".

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+For example, if you are running `--mainnet`, it will be
+`~/.lightning/bitcoin`.
+
+`hsm_secret`
+------------
```
Mix of `----` and `###` headings makes it more difficult to judge the overall structure. Please use `###` prefixes wherever possible.

ZmnSCPxj replied:
I mildly dislike the `###` prefixes since to me "more `#` = more heavier" i.e. `###` should map to `<h1>`, `#` should map to `<h3>`.  *shrug*.  Changed anyway.

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+or even to complete failure to recover, so use the other methods below
+first to recover!
+
+### Backing Up While `lightningd` Is Running
```
Ok, while the above should be the last option presented, this one should be even laster! :wink:

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+* ZFS RAID-Z, a filesystem that cannot be legally distributed with the Linux
+  kernel, but can be distributed in a BSD system, and can be installed 
+  on Linux with some extra effort, see
+  [ZFSonLinux](https:://zfsonlinux.org).
```
```suggestion
  [ZFSonLinux](https://zfsonlinux.org).
```

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+  This creates an initial copy of the database at the NFS mount.
+* Add these settings to your `lightningd` configuration:
+  * `important-plugin=/path/to/backup.py`
+  * `backup-destination=file:///path/to/nfs/mount`
```
This is no longer needed, because we now write a `lock` file into the `$LIGHTNINGDIR` in order to be functional right-away, and avoid accidentally changing the backup location.

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+`lightningd` may also be compiled with PostgreSQL support.
+PostgreSQL is generally faster than SQLITE3, and also supports running a
+PostgreSQL cluster to be used by `lightningd`, with automatic replication
+and turnover in case an entire node of the PostgreSQL cluster fails.
```
```suggestion
and failover in case an entire node of the PostgreSQL cluster fails.
```

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+
+Setting this up, however, is more involved.
+
+By default, `lightningd` compiles without PostgreSQL support.
```
We compile with postgresql by default if `./configure` detect `libpq`, and our releases are configured to ship with postgres enabled.

cdecker commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+
+The above guide assumes you are setting up a new node from scratch.
+It is also specific to PostgreSQL 12, and setting up for other versions
+**will** have differences; read the PostgreSQL manuals linked above.
```
I'm somehow missing the `--wallet-dsn=postgres://...` pointer to switch over to the postgres DB.

ZmnSCPxj replied:
It is `--wallet`, and it is mentioned in the gabridome guide, but will add a random reference to it here.

cdecker reviewed (commented):
Sorry for the delay reviewing this. It looks quite good, but reordering might be good, to emphasize which mechanisms are preferred, and for whom it might be worth it:

 - Backup `hsm_secret`: static, for all users
 - Backup plugins for end-users
 - Postgresql replication for enterprises
 - File-backup: if no other option, and with appropriate warnings
   - Backup while offline
   - Hot backups while the node is running

svewa commented:
Seeing the discussion on how to backup the sqlite file, would it not be best to not filesystem-copy it, but sqlite3 .dump it? https://www.sqlitetutorial.net/sqlite-dump/ or would this not guarantee consistency?

darosior commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,413 @@
+It is also specific to PostgreSQL 12, and setting up for other versions
+**will** have differences; read the PostgreSQL manuals linked above.
+
+If you want to continue a node that started using an SQLITE3 database,
+use the [tool by @fiatjaf][fiatjafmcldsp] to convert your SQLITE3 database
+to a PostgreSQL database.
```
Hmmm .. Are you confident with it ? Even @fiatjaf [does not seem to be](https://github.com/fiatjaf/mcldsp#beware-this-may-destroy-your-life-and-make-you-lose-all-your-money-use-at-your-own-risk).  If we are confident we are able to provide such a **crucial** tool it shoudl probably be part of the main repo...

ZmnSCPxj replied:
No, have never used it, will remove the link instead.

darosior reviewed (commented):
Really nice! I now have a reference to point people to instead of handwaving the different possibilities each time i'm asked :)

Just a small comment re the db migration tool which i'm not sure is safe.

darosior commented:
> Hi,
there's a new possibility for the backup of hsm_secret file.
I find it useful because it converts to xpriv format, so you can import into one descriptor based wallet in Bitcoin Core or wherever xpriv are accepted.
https://github.com/domegabri/lightning-secret

FWIW, it was already part of `hsmtool` since last release..
Descriptor are really neat but ya footgun in our case, and yea, as Zmn points out you cannot recover the `hsm_secret` from the xpriv.

ZmnSCPxj commented:
> Seeing the discussion on how to backup the sqlite file, would it not be best to not filesystem-copy it, but sqlite3 .dump it? https://www.sqlitetutorial.net/sqlite-dump/ or would this not guarantee consistency?

It locks the file while doing the dump.  The `VACUUM INTO` query is similar.  If `lightningd` accesses the database while your backup process is `.dump`ing or `VACUUM INTO`ing, it will crash `lightningd`, which is bad.

ZmnSCPxj commented:
Rebased, reordered sections as per @cdecker, changed to use `#` for section headings, various changes as per @cdecker , remove link to /fiatjaf/mcdlsp , warn against `.dump` and `VACUUM INTO`, mention use of multiple USB flash disks for RAID1.

darosior reviewed (approved):
ACK 948489ae08e735ea31c750a5bc68f6a82bd39e07
thanks for taking the time to write this up.

cdecker commented:
Excellent work @everyone, this was very much needed, and thanks in particular to @ZmnSCPxj for taking the initiative :+1:

ZmnSCPxj commented on doc/BACKUP.md:
```diff
@@ -0,0 +1,440 @@
+Even if you have one of the better options above, you might still want to do
+this as a worst-case fallback, as long as you:
+
+* Attempt to recover using the other backup options below first.
```
s/below/above/

LOL I did not fix it up after rearranging the text!

domegabri commented:
> > Hi,
> > there's a new possibility for the backup of hsm_secret file.
> > I find it useful because it converts to xpriv format, so you can import into one descriptor based wallet in Bitcoin Core or wherever xpriv are accepted.
> > https://github.com/domegabri/lightning-secret
> 
> FWIW, it was already part of `hsmtool` since last release (EDIT: and i noticed Riccardo Casatta did a new one last week too ! ) ..
> Descriptor are really neat but ya footgun in our case, and yea, as Zmn points out you cannot recover the `hsm_secret` from the xpriv.

Hi @darosior , I think the last release allows to dump public version of the descriptors. But I needed xpriv because I had a corrupt db and wanted to scan for coins and recover tehm

darosior commented:
@domegabri oh right we removed it in later stage as it was a nice footgun! https://github.com/ElementsProject/lightning/pull/4171#issue-513075486
